tools/codegen/api/cpp.py

```diff
@@ -272,46 +272,20 @@ def argument_faithful(
     else:
         return argument(a)
 
-# NB: this unconditionally groups arguments
 def group_arguments(
     func: FunctionSchema, *, method: bool
-) -> Sequence[Union[Argument, TensorOptionsArguments, ThisArgument]]:
-    args: List[Union[Argument, ThisArgument, TensorOptionsArguments]] = []
-
+) -> Sequence[Union[Argument, TensorOptionsArguments, SelfArgument]]:
+    args: List[Union[Argument, SelfArgument, TensorOptionsArguments]] = []
     args.extend(func.arguments.out)
-
-    if method:
-        args.extend(ThisArgument(a) if a.name == "self" else a for a in func.arguments.positional)
-    else:
-        args.extend(func.arguments.positional)
-
-    # group up arguments for tensor options
-
-    def pred(name: str, ty: Type) -> Callable[[Argument], bool]:
-        return lambda a: a.name == name and a.type in [ty, OptionalType(ty)]
-    predicates = [  # order matters
-        pred('dtype', Type.parse('ScalarType')),
-        pred('layout', Type.parse('Layout')),
-        pred('device', Type.parse('Device')),
-        pred('pin_memory', Type.parse('bool')),
-    ]
-
-    i = 0
-    while i < len(func.arguments.kwarg_only):
-        # If there is enough space...
-        if i <= len(func.arguments.kwarg_only) - len(predicates):
-            # And the next len(predicates) arguments look like TensorOptions arguments
-            if all(p(a) for p, a in zip(predicates, func.arguments.kwarg_only[i : i + len(predicates)])):
-                # Group them together as one argument
-                args.append(TensorOptionsArguments(
-                    dtype=func.arguments.kwarg_only[i],
-                    layout=func.arguments.kwarg_only[i + 1],
-                    device=func.arguments.kwarg_only[i + 2],
-                    pin_memory=func.arguments.kwarg_only[i + 3],
-                ))
-                i += len(predicates)
-                continue
-        args.append(func.arguments.kwarg_only[i])
-        i += 1
-
+    args.extend(func.arguments.pre_self_positional)
+    if func.arguments.self_arg is not None:
+        if method:
+            args.append(func.arguments.self_arg)
+        else:
+            args.append(func.arguments.self_arg.argument)
+    args.extend(func.arguments.post_self_positional)
+    args.extend(func.arguments.pre_tensor_options_kwarg_only)
+    if func.arguments.tensor_options is not None:
+        args.append(func.arguments.tensor_options)
+    args.extend(func.arguments.post_tensor_options_kwarg_only)
```

Comment:
This is another use of the fields, although the intention is to kill this function and make use of the fields directly